### PR TITLE
Add a flag for enabling experimental features

### LINKED ISSUE #2035: Add experimental feature framework to config system.
We want to have a global opt-in to enable experimental features, plus an 'experimental namespace' for specific feature parameters.

### PR DESCRIPTION
**Pull request by str4d**
Closes #2035.

### CODE REVIEW

**Comment by daira** [comment]
I'm assuming this should not be reviewed yet because it has [WIP].

**Comment by str4d** [comment]
@daira ready for review, now that https://github.com/zcash/zcash/issues/2035#issuecomment-277848556 is addressed.

**Comment by bitcartel** [comment]
utACK

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2002,7 +2002,7 @@ Value encryptwallet(const Array& params, bool fHelp)
     if (!EnsureWalletIsAvailable(fHelp))
         return Value::null;
 
-    auto fEnableWalletEncryption = GetBoolArg("-developerencryptwallet", false);
+    auto fEnableWalletEncryption = fExperimentalMode && GetBoolArg("-developerencryptwallet", false);
```
Comment (daira) [inline src/wallet/rpcwallet.cpp]: This silently ignores ``-developerencryptwallet`` when not in experimental mode. It should give an error.

Comment (daira) [inline src/wallet/rpcwallet.cpp]: Otherwise ut(ACK+cov).

**Comment by bitcartel** [comment]
@str4d Minor change requested in review from @daira

**Comment by str4d** [comment]
That shouldn't be the case: it should trigger the same RPC error as before, because the change solely adds an extra boolean restriction.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (str4d) [inline src/wallet/rpcwallet.cpp]: Ah, you mean there should be an error saying "please enable `-experimentalfeatures`".

**Comment by str4d** [comment]
@bitcartel @daira Amended commit to add an init check.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (daira) [inline src/wallet/rpcwallet.cpp]: Given the check [here](https://github.com/zcash/zcash/pull/2056/files#diff-c865a8939105e6350a50af02766291b7R757), this change is no longer required.

**Comment by bitcartel** [comment]
ACK

**Comment by bitcartel** [comment]
@zkbot r+

**Pull request merged**